torchtune/training/checkpointing/_checkpointer.py

```diff
@@ -912,3 +923,182 @@ def save_checkpoint(
+
+        load(
+            state_dict,
+            storage_reader=FileSystemReader(checkpoint_path),
+        )
+
+        return state_dict
+
+    def save_checkpoint(
+        self,
+        state_dict: Dict[str, Any],
+        epoch: int,
+        save_async: bool = False,
+    ) -> None:
+        """
+        Save a distributed checkpoint to disk.
+        If ``save_async`` is True, the save happens asynchronously unblocking the GPUs faster.
+
+        Args:
+            state_dict (Dict[str, Any]): Checkpoint state dict to be written out to file
+            epoch (int): Epoch number. Used to create the checkpoint file name
+            save_async (bool): If True, save the checkpoint asynchronously
+        """
+
+        if self._is_rank_zero:
+            logger.info(
+                f"DistributedCheckpointer is saving a checkpoint for the epoch {epoch}"
+            )
+
+        checkpoint_path = Path.joinpath(
+            self._output_dir, f"{self._checkpoint_dir_prefix}_{epoch}"
+        )
+
+        if self._checkpoint_future is not None:
+            # Previous checkpoint needs to finish before saving the next one.
+            wait_start = time.perf_counter()
+
+            if self._is_rank_zero:
+                logger.info(
+                    "Previous checkpoint has not finished. Checkpointing frequency is too high. Waiting..."
+                )
+
+            self._checkpoint_future.result()
+
+            if self._is_rank_zero:
+                logger.info(
+                    f"Waited {time.perf_counter() - wait_start:.2f} seconds for previous checkpoint to finish"
+                )
+
+        cp_start = time.perf_counter()
+
+        if save_async:
+
+            def callback(
+                f: Future,
+            ) -> None:
+                if f.exception() is None:
+                    if self._is_rank_zero:
+                        logger.info(
+                            f"Checkpoint is saved asynchronously to {checkpoint_path}"
+                        )
+
+                    for index in range(epoch):
+                        last_checkpoint = os.path.join(
+                            self._checkpoint_dir,
+                            f"{self._checkpoint_dir_prefix}_{index}",
+                        )
+                        if os.path.isfile(
+                            os.path.join(last_checkpoint, self._metadata_file)
+                        ):
+                            if self._is_rank_zero:
+                                logger.info(
+                                    f"Cleaning up the last checkpoint {last_checkpoint} to save storage"
+                                )
+                            shutil.rmtree(last_checkpoint)
+                else:
+                    logger.error(
+                        f"Checkpoint failed to save asynchronously to {checkpoint_path} with the exception {f.exception()}"
+                    )
```

Comment:
I see a bunch of exceptions coming from rmtree when I run this on my machine. E.g.

```
exception calling callback for <Future at 0x7f7415a06a50 state=finished returned Metadata>
Traceback (most recent call last):
  File "/home/ebs/.conda/envs/tt-alt-10-24/lib/python3.11/concurrent/futures/_base.py", line 340, in _invoke_callbacks
    callback(self)
  File "/data/users/ebs/ebs-torchtune-alt/torchtune/training/checkpointing/_checkpointer.py", line 1062, in callback
    shutil.rmtree(last_checkpoint)
  File "/home/ebs/.conda/envs/tt-alt-10-24/lib/python3.11/shutil.py", line 752, in rmtree
    _rmtree_safe_fd(fd, path, onerror)
  File "/home/ebs/.conda/envs/tt-alt-10-24/lib/python3.11/shutil.py", line 703, in _rmtree_safe_fd
    onerror(os.unlink, fullname, sys.exc_info())
  File "/home/ebs/.conda/envs/tt-alt-10-24/lib/python3.11/shutil.py", line 701, in _rmtree_safe_fd
    os.unlink(entry.name, dir_fd=topfd)
FileNotFoundError: [Errno 2] No such file or directory: '__2_46.distcp'
```

(However I am still able to save the checkpoints successfully)

Reply:
Ohh, my bad. Thanks for catching that. I somehow did not notice this in my testing. Basically the checkpoint deletion should happen at rank_0 only otherwise two different ranks may have a race condition in deleting a file and will run into the error above. Its not a problem during save, since DCP prepares the global plan and distributes it across ranks so every rank knows exactly what needs to be saved and there is no collision there.

Reply:
Since you suggested to remove this functionality for now from this PR, this logic will get cleaned up. I will record this issue it in my doc of suggestions for refactor and we can get this added for all Checkpointers in one go.